### DISCUSSION TITLE
soft avoid to hard avoids

### DISCUSSION THREAD

**Discussion by ADMUKHE** [valhalla/valhalla#4449] [General]
As it is clear from the code and documentation, value for soft avoids like "use_tolls" or "use_highways" can range from 0 to 1. However, 0 does not mean the feature would be definetely avoided since it could be necessary to generate routes. Is it possible to introduce hard avoids like "exclude_highways" or "exclude_tolls" which will definetely avoid the features and return "no route found" when its not possible to generate routes without this feature?

**Comment by nilsnolde** [comment]
Possible sure. Read the discussion in this PR: https://github.com/valhalla/valhalla/pull/3733, specifically https://github.com/valhalla/valhalla/pull/3733#issuecomment-1279060692.

**Comment by ADMUKHE** [comment]
inside search filters, there is also "exclude_tunnel" , "exclude_bridge", how is it different from the PR you mentioned above! Like functionally isint it meant to do the same thing?

**Reply by nilsnolde** [reply]
That’s a search filter to _locations_, not the expansion

**Reply by ADMUKHE** [reply]
yes i got the difference now! Thanks!
